File: packages/slot-math-kernels/src/slot_math_kernels/asymmetric_paytable.py
"""W244 wave 31 — closed-form analytical model for `asymmetric_paytable`.

Industry pattern (NetEnt Twin Spin asymmetric reels, Pragmatic Wild West
Gold asymmetric pays, ELK Wild Toro asymmetric stacks):

  Asymmetric paytable
  -------------------
    Per-symbol pay depends on PER-REEL appearance count, not just
    "k-of-a-kind anywhere on line". Specifically: standard line games
    assume a symbol's pay scales as f(k) regardless of which reel
    contributes; asymmetric games have per-reel weights so the EFFECTIVE
    pay distribution depends on which reels show the symbol.

    Simplest closed-form: operator supplies per-symbol-and-reel pay
    contribution table, kernel aggregates.

  Closed-form RTP contribution
  ----------------------------
    Per-spin RTP = sum over (symbol, reel-set) of
        P(reel-set ∈ winning pattern) × pay[symbol][reel-set-shape]

    Operator passes `per_shape_contribution[shape_key]` per symbol,
    sourced from PAR or MC. Kernel aggregates.

Pure-stdlib.
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class AsymmetricPaytableParams:
    """Operator-supplied per-symbol shape-key contributions.

    Each entry `per_symbol_contributions[symbol][shape_key]` is the
    per-spin RTP contribution × bet from that (symbol, shape) pair.
    """
    per_symbol_contributions: dict[str, dict[str, float]]

    def __post_init__(self):
        if not self.per_symbol_contributions:
            raise ValueError("per_symbol_contributions must be non-empty")
        for sym, table in self.per_symbol_contributions.items():
            if not table:
                raise ValueError(f"per_symbol_contributions[{sym!r}] must be non-empty")
            for shape, v in table.items():
                if v < 0:
                    raise ValueError(
                        f"contribution[{sym!r}][{shape!r}] = {v} must be ≥ 0"
                    )


def asymmetric_paytable_rtp(params: AsymmetricPaytableParams) -> dict:
    """Per-spin RTP + per-symbol breakdown."""
    total = 0.0
    per_symbol = []
    for sym in sorted(params.per_symbol_contributions.keys()):
        table = params.per_symbol_contributions[sym]
        sym_total = sum(table.values())
        total += sym_total
        per_symbol.append({
            "symbol": sym,
            "total_contribution_x_bet": sym_total,
            "per_shape": dict(table),
        })
    return {
        "rtp_contribution": total,
        "symbols_count": len(params.per_symbol_contributions),
        "per_symbol_breakdown": per_symbol,
    }

File: rust-sim/src/kernels/asymmetric_paytable.rs
//! W244.31 — asymmetric_paytable Rust port.
//!
//! Mirror of `tools/math_dsl/asymmetric_paytable.py`. Closed-form:
//!   RTP_total = Σ_symbol Σ_shape per_symbol_contributions[symbol][shape]
//!
//! Operator pre-computes per-(symbol, shape-key) contribution from PAR or
//! MC; kernel aggregates. Used by NetEnt Twin Spin, Yggdrasil Wild West
//! Gold, Microgaming Wild Toro (any game where the same symbol pays
//! differently depending on which reels it occupies).

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AsymmetricPaytableParams {
    /// {symbol → {shape_key → contribution_x_bet}}
    pub per_symbol_contributions: BTreeMap<String, BTreeMap<String, f64>>,
}

impl AsymmetricPaytableParams {
    pub fn validate(&self) -> Result<(), String> {
        if self.per_symbol_contributions.is_empty() {
            return Err("per_symbol_contributions must be non-empty".into());
        }
        for (sym, table) in &self.per_symbol_contributions {
            if table.is_empty() {
                return Err(format!(
                    "per_symbol_contributions[{}] must be non-empty",
                    sym
                ));
            }
            for (shape, &v) in table {
                if v < 0.0 {
                    return Err(format!(
                        "contribution[{}][{}] = {} must be ≥ 0",
                        sym, shape, v
                    ));
                }
            }
        }
        Ok(())
    }
}

#[derive(Debug, Serialize)]
pub struct PerSymbol {
    pub symbol: String,
    pub total_contribution_x_bet: f64,
    pub per_shape: BTreeMap<String, f64>,
}

#[derive(Debug, Serialize)]
pub struct AsymmetricPaytableResult {
    pub rtp_contribution: f64,
    pub symbols_count: usize,
    pub per_symbol_breakdown: Vec<PerSymbol>,
}

pub fn asymmetric_paytable_rtp(params: &AsymmetricPaytableParams) -> AsymmetricPaytableResult {
    let mut total = 0.0_f64;
    let mut per_symbol = Vec::with_capacity(params.per_symbol_contributions.len());
    // BTreeMap iterates sorted by key (matches Python `sorted(...)` semantics).
    for (sym, table) in &params.per_symbol_contributions {
        let sym_total: f64 = table.values().sum();
        total += sym_total;
        per_symbol.push(PerSymbol {
            symbol: sym.clone(),
            total_contribution_x_bet: sym_total,
            per_shape: table.clone(),
        });
    }
    AsymmetricPaytableResult {
        rtp_contribution: total,
        symbols_count: params.per_symbol_contributions.len(),
        per_symbol_breakdown: per_symbol,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn twin_spin_like() -> AsymmetricPaytableParams {
        let mut wd1 = BTreeMap::new();
        wd1.insert("twin_reels".to_string(), 0.20);
        wd1.insert("triple_reels".to_string(), 0.10);
        wd1.insert("quad_reels".to_string(), 0.05);

        let mut wd2 = BTreeMap::new();
        wd2.insert("twin_reels".to_string(), 0.10);
        wd2.insert("triple_reels".to_string(), 0.05);

        let mut top = BTreeMap::new();
        top.insert("S_HP1".to_string(), wd1);
        top.insert("S_HP2".to_string(), wd2);
        AsymmetricPaytableParams { per_symbol_contributions: top }
    }

    #[test]
    fn aggregates_correctly() {
        let p = twin_spin_like();
        let r = asymmetric_paytable_rtp(&p);
        // Σ all values: 0.20+0.10+0.05+0.10+0.05 = 0.50
        assert!((r.rtp_contribution - 0.50).abs() < 1e-12);
        assert_eq!(r.symbols_count, 2);
    }

    #[test]
    fn per_symbol_breakdown_sums_correctly() {
        let p = twin_spin_like();
        let r = asymmetric_paytable_rtp(&p);
        // First symbol (S_HP1 alphabetically) → 0.35
        let hp1 = r.per_symbol_breakdown.iter()
            .find(|s| s.symbol == "S_HP1").unwrap();
        assert!((hp1.total_contribution_x_bet - 0.35).abs() < 1e-12);
        // Second symbol (S_HP2) → 0.15
        let hp2 = r.per_symbol_breakdown.iter()
            .find(|s| s.symbol == "S_HP2").unwrap();
        assert!((hp2.total_contribution_x_bet - 0.15).abs() < 1e-12);
    }

    #[test]
    fn validate_rejects_empty() {
        let p = AsymmetricPaytableParams {
            per_symbol_contributions: BTreeMap::new(),
        };
        assert!(p.validate().is_err());
    }

    #[test]
    fn validate_rejects_negative_contribution() {
        let mut inner = BTreeMap::new();
        inner.insert("shape".to_string(), -0.5);
        let mut top = BTreeMap::new();
        top.insert("sym".to_string(), inner);
        let p = AsymmetricPaytableParams { per_symbol_contributions: top };
        assert!(p.validate().is_err());
    }
}
